tests/package_tests/docker_test.py

```diff
@@ -0,0 +1,119 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import argparse
+import pathlib as pl
+import subprocess
+import sys
+import time
+import datetime
+import os
+from typing import Union
+
+sys.path.append(str(pl.Path(__file__).absolute().parent.parent.parent))
+
+from tests.package_tests.common import PipeReader, check_agent_log_request_stats_in_line, assert_and_throw_if_line_an_error
+from tests.package_tests.common import COMMON_TIMEOUT
+
+
+def build_agent_image(builder_path: pl.Path):
+    # Call the image builder script.
+    subprocess.check_call(
+        [str(builder_path), "--tags", "docker-test"],
+    )
+
+
+_AGENT_CONTAINER_NAME = "scalyr-agent-docker-test"
+
+
+def _delete_agent_container():
+
+    # Kill and remove the previous container, if exists.
+    subprocess.check_call(
+        ["docker", "rm", "-f", _AGENT_CONTAINER_NAME]
+    )
+
+
+def _test(scalyr_api_key: str):
+
+    # Run agent inside the container.
+    subprocess.check_call(
+        [
+            "docker", "run", "-d", "--name", _AGENT_CONTAINER_NAME, "-e", f"SCALYR_API_KEY={scalyr_api_key}",
+            "-v", "/var/run/docker.sock:/var/scalyr/docker.sock", "scalyr/scalyr-agent-docker-json:docker-test"
+        ]
+    )
+
+    # Wait a little.
+    time.sleep(3)
+
+    # Execute tail -f command on the agent.log inside the container to read its content.
+    agent_log_tail_process = subprocess.Popen(
+        ["docker", "exec", "-i", _AGENT_CONTAINER_NAME, "tail", "-f", "-n+1", "/var/log/scalyr-agent-2/agent.log"],
+        stdout=subprocess.PIPE
+    )
+
+    # Read lines from agent.log. Create pipe reader to read lines from the previously created tail process.
+    # Also set the mode for the process' std descriptor as non-blocking.
+    os.set_blocking(agent_log_tail_process.stdout.fileno(), False)
+    pipe_reader = PipeReader(pipe=agent_log_tail_process.stdout)
+
+    timeout_time = datetime.datetime.now() + datetime.timedelta(seconds=COMMON_TIMEOUT)
+    try:
+        while True:
+
+            seconds_until_timeout = (timeout_time - datetime.datetime.now()).seconds
+            if seconds_until_timeout <= 0:
+                raise TimeoutError("Test has timed out.")
+            line = pipe_reader.next_line(timeout=seconds_until_timeout)
+            # Look for any ERROR message.
+            assert_and_throw_if_line_an_error(line)
+
+            # TODO: add more checks.
+
+            if check_agent_log_request_stats_in_line(line):
```

Comment:
To make it easier to troubleshoot (especially scenarios when there is no data for a while), it would probably be good to log something during each poll interval - e.g. "Polling pipe for new line and waiting for x seconds".